pytorch_lightning/loops/base.py

```diff
@@ -45,8 +47,35 @@ class Loop(ABC):
 
     def __init__(self) -> None:
         self.iteration_count: int = 0
-        self.trainer: Optional['pl.Trainer'] = None
         self.restarting = False
+        self._trainer: Optional['pl.Trainer'] = None
+
+    @property
+    def loop_progress(self) -> Dict[str, Any]:
+        """Return the progress for the current loop and children loop."""
+        progress = {}
+        for k, v in self.__dict__.items():
+            if isinstance(v, BaseProgress):
+                progress[k] = v
+            elif isinstance(v, Loop):
+                progress[k] = v.loop_progress
+        return progress
+
+    @property
+    def trainer(self) -> Optional['pl.Trainer']:
+        return self._trainer
+
+    @trainer.setter
+    def trainer(self, trainer: 'pl.Trainer'):
+        """Connect the Trainer to itself and all its children loops"""
+        if not isinstance(trainer, pl.Trainer):
+            raise MisconfigurationException(
+                f"Loop {self.__class__.__name__} should be connected to a `Trainer`, found: {trainer}."
+            )
+        self._trainer = trainer
+        for v in self.__dict__.values():
```

Comment:
should we do the assignment recursively? I would actually prefer to do this more explicitly in the trainer (and also checking the loops properties maybe).

cc @awaelchli

Reply:
I think we need such a mechanism  one way or the other and the trainer assignment can't go through the connect() method if connect() is the interface the user has to link loops together. It's more flexible for the user not to have to create a trainer instance first to connect custom loops. Plus it's safer imo to not allow a user to set different trainer references for loops.

Reply:
closing this one